### PR TITLE
Switch Travis to Trusty

### PR DESCRIPTION
**Pull request by theuni**
I worked with Travis to get the cache flag on my personal repo for testing, and I'm happy to report that all works great (after a day of churning builds).

For some background: Travis has enabled caching for their new infrastructure, but the feature is brand new and not yet generally available. For now it requires a special flag (set on their end) on your repo. Once they're satisfied with its stability, the flag will be granted to everyone. Note that the flag comes with the caveat of invalidating all current caches. That's not a problem for us, other than slowing things down for a few hours. I was primarily worried about unintended side-effects of opting in, so I added it on my own repo for testing before requesting it here.

If there's no reason to delay our move, I'll request the flag for this repo tomorrow (4/21). 

Once that's take care of, we can merge this for Trusty goodness. On the day that Xenial is scheduled to be released :p.

Next step is c++11, which should be painless to enable now.

Note that this includes #7792, otherwise the OSX build fails. @laanwj I'm happy to rebase if you'd prefer to merge that first.

### CODE REVIEW

**Comment by theuni** [comment]
Note that the build failed because we currently have a (different) special flag on our repo that force-routes us to the legacy infrastructure for caching. This change will disable that in favor of the new testing flag.

(It may sound convoluted, but this change moves us one step closer towards being a normal repo. Once the caching feature is out of beta, there will be nothing special about us.)

File: .travis.yml [hunk 1]
```diff
@@ -38,22 +37,25 @@ matrix:
...
     - compiler: ": Win64"
-      env: HOST=x86_64-w64-mingw32 PPA="ppa:ubuntu-wine/ppa" PACKAGES="nsis gcc-mingw-w64-x86-64 g++-mingw-w64-x86-64 binutils-mingw-w64-x86-64 mingw-w64-dev wine1.7 bc" RUN_TESTS=true GOAL="deploy" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports" MAKEJOBS="-j2"
+      env: HOST=x86_64-w64-mingw32 DPKG_ADD_ARCH="i386" PACKAGES="nsis g++-mingw-w64-x86-64 wine1.6 bc" RUN_TESTS=true GOAL="deploy" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports" MAKEJOBS="-j2"
     - compiler: ": bitcoind"
-      env: HOST=x86_64-unknown-linux-gnu PACKAGES="bc python-zmq" PPA="ppa:chris-lea/zeromq" DEP_OPTS="NO_QT=1 NO_UPNP=1 DEBUG=1" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --enable-glibc-back-compat --enable-reduce-exports CPPFLAGS=-DDEBUG_LOCKORDER"
+      env: HOST=x86_64-unknown-linux-gnu PACKAGES="bc python-zmq python3-zmq" DEP_OPTS="NO_QT=1 NO_UPNP=1 DEBUG=1" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --enable-glibc-back-compat --enable-reduce-exports CPPFLAGS=-DDEBUG_LOCKORDER"
```
Comment (MarcoFalke) [inline .travis.yml]: Mind to share why the py2 and py3 packages of zmq are needed?

**Comment by MarcoFalke** [comment]
Awesome! Concept ACK!

**Comment by theuni** [comment]
@MarcoFalke Just for the sake of being future-proof. It only uses the python2 one at the moment, but I believe the (your?) intention is to move the rpc tests to python3. Just doing python2 for now is fine by me if that's what you'd prefer. I was just saving some future head-scratching when builds fail after the switch.

File: .travis.yml (hunk 1 above)
Comment (MarcoFalke) [inline .travis.yml]: @theuni  Jup, let's keep it py2 only for now, if possible.

Comment (theuni) [inline .travis.yml]: Sure, will remove.

**Comment by gmaxwell** [comment]
Thanks for working on this.

**Comment by theuni** [comment]
Dropped the python3-zmq packages.

**Comment by theuni** [comment]
This was delayed because there's a little extra work needed in switching away from our old setup. ETA given from Travis is Monday.

**Comment by theuni** [comment]
Ok, finally all green and cache-enabled.

I disabled a few qt builds because they seem to take longer on this infrastructure than before. We can investigate and turn them back on asap.

For now, until this is merged, PRs will be unusually slow, so it's probably best to go ahead with this now and work through the regressions.

**Comment by theuni** [comment]
To expand on the "PRs will be unusually slow" bit above:

In removing the special bits from our repo, all caches have been invalidated. In theory, precise caches will be regenerated for subsequent PRs, but I'm nervous about that actually happening seamlessly on the deprecated platform.

Even if it does work without issue, PRs will be slow until _something_ is pushed into master. When a PR is created, the build cache is taken from its source branch (master), and copied into the PR. From there, the PR updates its own cache with each update.

Right now we have no master cache, and no PR caches. So until there's something to pull from master, each PR will be rebuilding from scratch. Also, once we _do_ pull the trigger on this, it will essentially invalidate all caches again, because all dependencies will be rebuilt for the new toolchains.

tl;dr: Need to push _something_ into master asap to avoid slow builds all day. May as well push this one, to avoid a repeat.

**Comment by MarcoFalke** [comment]
ACK on merging today.

**Comment by btcdrak** [comment]
ACK

**Comment by jonasschnelli** [comment]
ACK 174023c9b008fc02316bce972b0c1031de3feee3

**Comment by btcdrak** [comment]
@theuni Does this have to be backported as well?

**Comment by theuni** [comment]
@btcdrak Yes, the same issues apply to the other branches.
